### PR TITLE
improved task middleware to show skipped path

### PR DESCRIPTION
**Pull request by sio4**
Hi, as described as issue #1379, command `buffalo task middleware` does just show middleware stacks per `App` registered with `App.Group()` or `App.Resource()` and we cannot find which middleware was skipped (by `App.Middleware.Skip()`) in specific paths.

By following routes and check related app, this commit allows user can find which is real middleware stack for each path. Example output is like this:

```console
$ buffalo t middleware
-> /
   github.com/gobuffalo/buffalo.*App.defaultErrorMiddleware
   github.com/gobuffalo/buffalo.*App.PanicHandler
   github.com/gobuffalo/buffalo.RequestLoggerFunc
   github.com/gobuffalo/buffalo.sessionSaver
   github.com/gobuffalo/mw-forcessl.Middleware.func1
   github.com/gobuffalo/mw-paramlogger.ParameterLogger
   github.com/gobuffalo/mw-csrf.glob..func1
   github.com/gobuffalo/buffalo-pop/pop/popmw.Transaction.func2
   github.com/gobuffalo/mw-i18n.*Translator.Middleware.func1
   github.com/hyeoncheon/skel/actions.AuthorizeHandler
-> GET / (by actions.HomeHandler)
   github.com/gobuffalo/buffalo.*App.defaultErrorMiddleware
   github.com/gobuffalo/buffalo.*App.PanicHandler
   github.com/gobuffalo/buffalo.RequestLoggerFunc
   github.com/gobuffalo/buffalo.sessionSaver
   github.com/gobuffalo/mw-forcessl.Middleware.func1
   github.com/gobuffalo/mw-paramlogger.ParameterLogger
   github.com/gobuffalo/mw-csrf.glob..func1
   github.com/gobuffalo/buffalo-pop/pop/popmw.Transaction.func2
   github.com/gobuffalo/mw-i18n.*Translator.Middleware.func1
-> /api/v1
   github.com/gobuffalo/mw-csrf.glob..func1
-> GET /api/v1/ (by actions.HomeHandler)
   [none]
-> /api/v1/app/users has same middleware stack as /api/v1
-> /api/v1/users has same middleware stack as /api/v1
-> /auth
   github.com/gobuffalo/buffalo.*App.defaultErrorMiddleware
   github.com/gobuffalo/buffalo.*App.PanicHandler
   github.com/gobuffalo/buffalo.RequestLoggerFunc
   github.com/gobuffalo/buffalo.sessionSaver
   github.com/gobuffalo/mw-forcessl.Middleware.func1
   github.com/gobuffalo/mw-paramlogger.ParameterLogger
   github.com/gobuffalo/mw-csrf.glob..func1
   github.com/gobuffalo/buffalo-pop/pop/popmw.Transaction.func2
   github.com/gobuffalo/mw-i18n.*Translator.Middleware.func1
-> /users has same middleware stack as /
$ 
```

For handler which has skipped middleware, it shows like this:

```console
-> GET / (by actions.HomeHandler)
<...>
```

Stack output format is same as normal `App` but the title line contains method and handler name.

When there is no middleware for some handler or `App`, by `Clear()`, it shows "[none]".

```console
-> GET /api/v1/ (by actions.HomeHandler)
   [none]
```

In case of the `App` that have same middleware stack as its parent, it shows like this:

```console
-> /api/v1/app/users has same middleware stack as /api/v1
```

Please check this patch.

### CODE REVIEW

File: grifts.go [hunk 1]
```diff
@@ -49,11 +49,63 @@ func middlewareGrift(a *App) {
...
+		}
+	}
+	return nil
+}
+
+func printMiddlewareStackWithIndent(s string) {
+	if s == "" {
+		s = "[none]"
 	}
+	s = strings.Replace(s, "\n", "\n   ", -1)
```
Comment (markbates) [inline grifts.go]:
```suggestion
	s = strings.Replace(s, "\n", "\n\t", -1)
```

File: grifts.go [hunk 2]
```diff
@@ -49,11 +49,63 @@ func middlewareGrift(a *App) {
...
+	fmt.Printf("   %v\n", strings.TrimSpace(s))
```
Comment (markbates) [inline grifts.go]:
```suggestion
	fmt.Printf("\t%v\n", strings.TrimSpace(s))
```

File: grifts.go [hunk 3]
```diff
@@ -49,11 +49,63 @@ func middlewareGrift(a *App) {
...
+			if parent := getParentApp(r.App.root, r.App.Name); parent != nil {
+				pname = parent.Name
+			}
+
+			mws[r.App.Name] = r.App.Middleware.String()
+			if mws[pname] != mws[r.App.Name] {
+				fmt.Printf("-> %s\n", r.App.Name)
+				printMiddlewareStackWithIndent(mws[r.App.Name])
+			} else {
+				fmt.Printf("-> %s has same middleware stack as %v\n", r.App.Name, pname)
```
Comment (markbates) [inline grifts.go]:
```suggestion
				fmt.Printf("-> %s (see: %v)\n", r.App.Name, pname)
```

File: grifts.go (hunk 1 above)
Comment (sio4) [reply grifts.go]: OK, I agree with using tab even though it takes five more spaces and can generate line wrap on 80 col terminal like me, anyway that is fancier!

**Comment by sio4** [comment]
Anyway, I found a bug, in case of the handler `Skip()`ed more than two middlewares! At the first time, I think about it but I missed while I focused on how to find the parent of the `App` since it only has `root`.

I will fix it as soon as possible, so please wait an hour! :-)

**Comment by markbates** [comment]
Awesome. Let me know when it’s ready to go.

**Comment by sio4** [comment]
OK, two commits for suggestion and bugfix are pushed and checked for CIs.
Please check this change. Now it just shows us single correct middleware stack for skipped, skipped handlers.

**Comment by sio4** [comment]
Hi @markbates,
Did you check updated commits? Just as reminder.

**Comment by sio4** [comment]
You're welcome!
Thanks for great framework everyday!

**Pull request merged**